daft/execution/physical_plan.py

```diff
@@ -263,43 +283,67 @@ def actor_pool_project(
         num_gpus=resource_request.num_gpus, memory_bytes=resource_request.memory_bytes
     )
 
-    with actor_pool_manager.actor_pool_context(
-        actor_pool_name,
-        actor_resource_request,
-        task_resource_request,
-        num_actors,
-        projection,
-    ) as actor_pool_id:
-        child_plan_exhausted = False
+    actor_pool_id = None
+
+    child_plan_exhausted = False
 
+    try:
         # Loop until the child plan is exhausted and there is no more work in the pipeline
         while not (child_plan_exhausted and len(child_materializations) == 0 and len(actor_pool_materializations) == 0):
-            # Exhaustively pop ready child_steps and submit them to be run on the actor_pool
-            while len(child_materializations) > 0 and child_materializations[0].done():
-                next_ready_child = child_materializations.popleft()
-                actor_project_step = (
-                    PartitionTaskBuilder[PartitionT](
-                        inputs=[next_ready_child.partition()],
-                        partial_metadatas=[next_ready_child.partition_metadata()],
-                        actor_pool_id=actor_pool_id,
-                    )
-                    .add_instruction(
-                        instruction=execution_step.StatefulUDFProject(projection),
-                        resource_request=task_resource_request,
+            # Pop ready child_steps and submit them to be run on the actor_pool
+            in_progress_child_materializations = list()
+            for next_child in child_materializations:
+                if next_child.task.done():
+                    # Initialize actor pool upon first task that is ready to be submitted
+                    if actor_pool_id is None:
+                        actor_pool_id = actor_pool_manager.setup_actor_pool(
+                            actor_pool_name,
+                            actor_resource_request,
+                            task_resource_request,
+                            num_actors,
+                            projection,
+                        )
```

Comment:
I see what you're doing.... I think the idea here is good but maybe we can find a better abstraction for lazy initialization of the actor pool.

WDYT of:

```python
with actor_pool_manager.actor_pool_context(...) as lazy_actor_pool:
    ...
    actor_pool_id = lazy_actor_pool.get_or_create()
```

I think this should let us have the same effect but have fewer invariants around local state of `None` or not `None`